README.md
# ModifiedParsing
Collection of scripts and programs related to the augmented blockParse.py build.


blockParseFinal.py-- The modified blockParse script. Allows the replacement of gc check with just
G-check as well as the option to scan the reverse-compliment of candidates as well.

ParseAll.py-- Optimization program that can be applied for chromosomes in which .bed files have been 
created for G-percents of 0-25.

FileWriter.py-- Used to construct big combined probe library for G-percent in range 0-25.

scriptwriter.py-- Generates shell scripts for blockParseFinal parsing.

ProbeWorker.py-- Program for working with parsed probes dataframe using pandas.


changes.txt
Change Log between blockParse.py and blockParseFinal.py

Note: Most changes involved adding in checks for g-percent where normally
      gc-percent is checked. Beyond that changes involve adding in a reverse
      compliment check as well as adding a '+' or '-' to the .fastq depending
      on said check.


Changes related to G-percent Check:
Lines 71-72: Added values for needed for gpercent check, similar to gcpercent check
Line 98: Added value numG on a strand, similar to numGC
Lines 177-245: Added various operations and checks mirroring those for gc
Lines 248-364: Again mirrored operations for gc, simply subbing in plain g values
Lines 378-381: Added gCheck definition using above values
Line 478: Indicates the use of gCheck definition
Lines 581-621: Error windows for gCheck

 


Changes related to Reverse Compliment Check:
Line 116: Tracks reverse compliment check status
Line 119: Similar check but for added .fastq column
Line 121: Another similar check for writing the reverse sequence to .fastq
Lines 383-393: Checks reverse strand for G-percent
Line 487: Checks reverse strand if indicated along with gcheck
Line 493: Indicates to write '-' to .fastq file
Line 495: Inidcates to write reverse compliment to .fastq file
Lines 735-758: Various uses of the rPass value to indicate which strand to 
               print on .fastq--more in comments on section.
Lines 789-794: Writing the reverse compliment to file

Argument Changes:
-RC turns on reverse compliment check
-GON turns on gCheck instead of gcCheck
-g/-G are min and max G allowed respecitvely
-gc/-GC are min and max GC allowed respecitvely

For details on how exactly checks are implemented, see comments in relevant section of program


FileWriter.py
import sys

class FileWriter:
    
    def __init__(self):
        self.chrNum = str(sys.argv[1])
        pass

    def writeBigFile(self):
        bigList = self.createBigList(0,180000000)
        self.addGPercent(bigList)
        filename = "BigProbeFile" + self.chrNum + ".txt"
        outputFile = open(filename, "w")
        for probe in bigList:
            thing = str(probe[1]) + " " + str(probe[2]) + " " + str(probe[4]) + " " + str(probe[6])
            outputFile.write(thing)
            outputFile.write("\n")
        outputFile.close()

    def shortenList(self, start, stop, file):
        
        bigList = open(file, "r")

        finalList = []

        for line in bigList:
            
            tempList = line.split()

            tempStart = int(tempList[1])
            tempStop = int(tempList[2])

            
            if tempStart >= start and tempStop <= stop:
                finalList.append(line)

        finalFinalList = []
        for line in finalList:
            newList = line.split()
            finalFinalList.append(newList)
        
        bigList.close()
        return finalFinalList
    
    def addGPercent(self, noGList):
        probesParsed = 0

        for stuff in noGList:
            currSeq = stuff[4]
            total = len(currSeq)
            numG = 0.0
            for letter in currSeq:
                if letter == 'G':
                    numG = numG + 1.0
            
            percentG = numG / total
            stuff.append(percentG)
            probesParsed = probesParsed + 1
            if probesParsed % 10000 == 0:
                print str(probesParsed) + " out of " + str(len(noGList)) + " parsed"

    
    def createBigList(self, int1, int2):
        biggerList =[]
        for x in range(0, 26):
            fileName = "chr" + self.chrNum + "_" + str(x) + ".bed"
            shortenedList = self.shortenList(int1, int2, fileName)
            for probe in shortenedList:
                biggerList.append(probe)
            print fileName + " is shortened to probes between " + str(int1) + " and " + str(int2)
        return biggerList


writerThing = FileWriter()
writerThing.writeBigFile()





ProbeWorker.py
import pandas as pd
import ParseAllFinal as paf

toParse = paf.ParseAllFinal()


## Parses through list to optimize probes in given range
## Outputs as a csv file
def dataFrameCreator(toParse, text, start, end, saved, bigFile):
    print("Welcome to the data frame creator\n")
    dataName = saved
    fileName = bigFile
    allList = toParse.createBigList(start, end, text, fileName)
    parsedProbes = toParse.parseThrough(allList, start, end, text)
    dataFrame = pd.DataFrame(parsedProbes,columns=['Start','End','Strand','gPercent'],dtype=float)
    dataFrame['Start'] = dataFrame['Start'].astype(int)
    dataFrame['End'] = dataFrame['End'].astype(int)
    dataFrame['Strand'] = dataFrame['Strand'].astype(str)
    print(dataFrame)
    dataFrame.to_csv(dataName)


## Client part to choose option
print("Welcome to the Probe Worker")

choice = input("Enter 1 for parsing, 2 for narrowing: ")

if choice == "1":
    bigFile = input("Enter big file name: ")

    text = False
    textI = input("Show text information (y/n)? ")
    if textI == "y":
        text = True
    elif textI == "n":
        text = False
    else:
        raise ValueError("Enter 'y' or 'n'")


    start = int(input("Enter start coordinate: "))
    end = int(input("Enter end coordinate: "))

    if end < start:
        raise ValueError("Start coordinate must be less than end coordinate")

    name = input("Enter name for output csv file: ")

    dataFrameCreator(toParse, text, start, end, name, bigFile)

## Allows editing of csv file based on coordinate or g-percent
## Outputs as new csv file
elif choice == "2":
    fileName = input("Enter csv file: ")
    newFileName = input("Enter output file: ")
    df = pd.read_csv(fileName)

    start = input("Enter start coordinate: ")
    end = input("Enter end coordinate: ")
    gPercentLow = input("Enter low gPercent: ")
    gPercentHigh = input("Enter high gPercent: ")

    newSet = []

    for index, row in df.iterrows():
        if int(row["Start"]) >= int(start) and int(row["End"]) <= int(end):
            if float(row["gPercent"]) >= float(gPercentLow) and float(row["gPercent"]) <= float(gPercentHigh):
                newSet.append([int(row["Start"]), int(row["End"]), str(row["Strand"]), float(row["gPercent"])])

    dataFrame = pd.DataFrame(newSet,columns=['Start','End','Strand','gPercent'],dtype=float)
    dataFrame['Start'] = dataFrame['Start'].astype(int)
    dataFrame['End'] = dataFrame['End'].astype(int)
    dataFrame['Strand'] = dataFrame['Strand'].astype(str)

    print(dataFrame)
    dataFrame.to_csv(newFileName)

else:
    raise ValueError("Enter '1' or '2'")













ParseAllpy.py
## Parses through a collection of .bed files for a given chromosome
## and optimizes a probeset based off of lowest G-percent, while
## maintaing no overlap.

import sys

## Note to use this include the chromosome number as the only argument
## in command line


class ParseAll:
    
    def __init__(self):
        pass
    
    ## Included for those who only want to scan certain coordinates 
    ## Not implmented in normal usage below
    def shortenList(self, start, stop, file):
        
        bigList = open(file, "r")

        finalList = []

        for line in bigList:
            
            tempList = line.split()

            tempStart = int(tempList[1])
            tempStop = int(tempList[2])

            
            if tempStart >= start and tempStop <= stop:
                finalList.append(line)

        finalFinalList = []
        for line in finalList:
            newList = line.split()
            finalFinalList.append(newList)
        
        return finalFinalList
    
    ## Adds g-percent value to lists since it is not in .bed file
    def addGPercent(self, noGList):
        probesParsed = 0

        for stuff in noGList:
            currSeq = stuff[4]
            total = len(currSeq)
            numG = 0.0
            for letter in currSeq:
                if letter == 'G':
                    numG = numG + 1.0
            
            percentG = numG / total
            stuff.append(percentG)
            probesParsed = probesParsed + 1
            if probesParsed % 10000 == 0:
                print str(probesParsed) + " out of " + str(len(noGList)) + " parsed"
        print "Done calculating G-Percent"
        print "\n"


    ## Combines all BigProbeFiles
    def createBigList(self, int1, int2):
        biggerList =[]
        for x in range(0, 26):
            fileName = "chrX_chm13_RC" + str(x) + ".bed"
            shortenedList = self.shortenList(int1, int2, fileName)
            for probe in shortenedList:
                biggerList.append(probe)
        return biggerList

    ## Parses through Big List
    def parseThrough(self, finalProbeSet, int1, int2):

        coordinateSet = [] ## Set of all coordinates that are in the final probe set (so far)
        bestProbes = []
        totalProbes = len(finalProbeSet)
        probesParsed = 0

        for probe in finalProbeSet:
            start = int(probe[1])
            end = int(probe[2])
            noOverlap = True
            currNum = start

            while (currNum <= end) and noOverlap:
                if currNum in coordinateSet:
                    noOverlap = False
                else:
                    coordinateSet.append(currNum)
                    currNum = currNum + 1
            
            if noOverlap:
                bestProbes.append(probe)
            
            probesParsed = probesParsed + 1
            
        ## Some optional values that can be implemented if needed
        percentCoverage = ((len(coordinateSet) * 1.0) / (int2-int1)) * 100
        kb = ((int2 - int1) * 1.0) / 1000
        probesKB = len(bestProbes) / kb
        ## Final parsed probe set
        return bestProbes

    
    def checkNoOverlap(self, list):
    ## ONLY DO THIS IF THE LIST HAS BEEN SORTED BY POSITION FIRST
        x = 0
        onlyPositive = True
        while x < len(parsedProbes) - 1 and onlyPositive:
            big = int(parsedProbes[x+1][1])
            small = int(parsedProbes[x][2])
            if (big - small) <= 0:
                onlyPositive = False
            x = x + 1
        if onlyPositive:
            print "There are no overlaps"
        else:
            print "Some overlaps exist"
        print "\n"


toParse = ParseAll()
chrNum = str(sys.argv[1])

## BigProbeFiles can be gotten from nexus library
fileName = 'BigProbeFile' + chrNum + '.txt'
shortFileName = 'chr' + chrNum
fileThing = open(fileName, 'r')
finalList = []

## Turning BigProbeFile into lists readable by above methods
for line in fileThing:
    tempFinalList = []
    tempFinalList.append(shortFileName)
    tempList = line.split()
    for item in tempList:
        tempFinalList.append(item)
    finalList.append(tempFinalList)
    
fileThing.close()
parsedProbes = toParse.parseThrough(finalList, 0, 1800000000) ## Large length covers hg38

## Sorting parsed probes by length
parsedProbes.sort(key = lambda x: x[1])
testFile = open('Parsed22.txt', 'w')
for thing in parsedProbes:
    for item in thing:
        testFile.write(str(item) + ' ')
    testFile.write('\n')
testFile.close()

##toParse.checkNoOverlap(parsedProbes)





